code/modules/mapexporting/mapexporter.dm

```diff
@@ -0,0 +1,169 @@
+			//Stuff to add
+			else
+				location = get_area(place).type
+				objects = place
+				place = place.type
+			//====Saving shuttles only / non shuttles only====
+			var/is_shuttle_area = istype(location, /area/shuttle)
+			if((is_shuttle_area && shuttle_area_flag == SAVE_SHUTTLEAREA_IGNORE) || (!is_shuttle_area && shuttle_area_flag == SAVE_SHUTTLEAREA_ONLY))
+				place = /turf/template_noop
+				location = /area/template_noop
+				objects = list()
+			//====For toggling not saving areas and turfs====
+			if(!(save_flag & SAVE_AREAS))
+				location = /area/template_noop
+			if(!(save_flag & SAVE_TURFS))
+				place = /turf/template_noop
+			//====Generate Header Character====
+			var/header_char = calculate_tgm_header_index(y + (x * width), layers)	//The characters of the header
+			var/current_header = "(\n"										//The actual stuff inside the header
+			//Add objects to the header file
+			var/empty = TRUE
+			//====SAVING OBJECTS====
+			if(save_flag & SAVE_OBJECTS)
+				for(var/obj/thing in objects)
+					var/metadata = generate_tgm_metadata(thing, vars_to_save)
+					current_header += "[empty?"":",\n"][thing.type][metadata]"
+					empty = FALSE
+					//====SAVING SPECIAL DATA====
+					//This is what causes lockers and machines to save stuff inside of them
+					if(save_flag & SAVE_OBJECT_PROPERTIES)
+						var/custom_data = thing.on_object_saved()
+						current_header += "[custom_data ? ",\n[custom_data]" : ""]"
+			//====SAVING MOBS====
+			if(save_flag & SAVE_MOBS)
+				for(var/mob/living/thing in objects)
+					if(istype(thing, /mob/living/carbon))		//Ignore people, but not animals
+						continue
+					var/metadata = generate_tgm_metadata(thing, vars_to_save)
+					current_header += "[empty?"":",\n"][thing.type][metadata]"
+					empty = FALSE
+			current_header += "[empty?"":",\n"][place],\n[location])\n"
+			//====Fill the contents file====
+			//Compression is done here
+			var/position_of_header = header_dat.Find(current_header)
+			if(position_of_header)
+				//If the header has already been saved, change the character to the other saved header
+				header_char = header_chars[position_of_header]
+			else
+				header += "\"[header_char]\" = [current_header]"
+				header_chars += header_char
+				header_dat += current_header
+			contents += "[header_char]\n"
+		contents += "\"}"
+	return "[header][contents]"
+
+//Sorts maps in terms of their positions, so scrambled / odd shaped maps can be saved
+/proc/sort_map(var/list/map, minx, miny, maxx, maxy)
+	var/width = maxx - minx + 1
+	var/height = maxy - miny + 1
+	var/allTurfs = new/list(width, height)
+	for(var/turf/place in map)
+		allTurfs[place.x - minx + 1][place.y - miny + 1] = place
+	return allTurfs
+
+//vars_to_save = list() to save all vars
+/proc/generate_tgm_metadata(var/atom/O, var/list/vars_to_save = list("pixel_x", "pixel_y", "dir", "name", "req_access", "req_access_txt", "piping_layer", "color", "icon_state", "pipe_color", "amount"))
```

Comment:
There should be an easy way to modify the `vars_to_save` on a per-type basis. Have a look at how World Server does it with their `vars_to_save()` proc (IIRC that's the name of it).

Reply:
There is now an empty variable on every atom called vars_to_save, if the variable isn't set to anything the default vars to save will be saved (the one passed into the function), else the vars to save of that atom will be saved.

There is also a flag that works on objects and mobs called DONT_SAVE